Build with clang 18

jdblischak opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [x] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->

---

Test PR to build osx-* binaries with clang 18

* conda-forge updated from clang 17 to clang 18 a few days ago (Oct 30th) (https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6571)
* our nightly osx-arm64 feedstock builds failed that same night (reported the morning of Oct 31st) (https://github.com/TileDB-Inc/conda-forge-nightly-controller/issues/150)

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

jdblischak commented:
Rebased onto main to include https://github.com/conda-forge/tiledb-feedstock/pull/370 which installs capnproto from conda-forge, which should bypass the compilation error with clang 18

jdblischak commented:
Making progress.

Previously it failed during the configuration step when trying to compile capnproto ([link to old build log that installed capnproto via vcpkg](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1069763&view=logs&j=d6b58996-039f-5e48-56bf-c3a016e5cd7f&t=dd6a23d9-2356-5264-64bf-579875a5d8d6)):

```
Installing 17/19 capnproto:arm64-osx...
Building capnproto:arm64-osx...
-- [OVERLAY] Loading triplet configuration from: $SRC_DIR/triplets/arm64-osx.cmake
-- Installing port from location: $SRC_DIR/build/_deps/vcpkg-src/buildtrees/versioning_/versions/capnproto/4a615543c6406b84fc52a931335d7fdb70037627
-- Downloading https://github.com/capnproto/capnproto/archive/v1.0.1.tar.gz -> capnproto-capnproto-v1.0.1.tar.gz...
-- Extracting source $SRC_DIR/build/_deps/vcpkg-src/downloads/capnproto-capnproto-v1.0.1.tar.gz
-- Applying patch disable-C-20-co-routines.patch
-- Applying patch undef-KJ_USE_EPOLL-for-ANDROID_PLATFORM-23.patch
-- Using source at $SRC_DIR/build/_deps/vcpkg-src/buildtrees/capnproto/src/v1.0.1-21327bcba6.clean
-- Configuring arm64-osx
CMake Error at scripts/cmake/vcpkg_execute_required_process.cmake:112 (message):
    Command failed: $SRC_DIR/build/_deps/vcpkg-src/downloads/tools/ninja/1.10.2-osx/ninja -v
    Working Directory: $SRC_DIR/build/_deps/vcpkg-src/buildtrees/capnproto/arm64-osx-rel/vcpkg-parallel-configure
    Error code: 1
    See logs for more information:
      $SRC_DIR/build/_deps/vcpkg-src/buildtrees/capnproto/config-arm64-osx-rel-CMakeCache.txt.log
      $SRC_DIR/build/_deps/vcpkg-src/buildtrees/capnproto/config-arm64-osx-out.log

Call Stack (most recent call first):
  $SRC_DIR/build/vcpkg_installed/x64-osx/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:252 (vcpkg_execute_required_process)
  buildtrees/versioning_/versions/capnproto/4a615543c6406b84fc52a931335d7fdb70037627/portfile.cmake:19 (vcpkg_cmake_configure)
  scripts/ports.cmake:170 (include)


error: building capnproto:arm64-osx failed with: BUILD_FAILED
```

Now that capnproto is installed from conda-forge, the configuration step completed successfully ([latest build log](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=1077337&view=logs&j=d6b58996-039f-5e48-56bf-c3a016e5cd7f&t=dd6a23d9-2356-5264-64bf-579875a5d8d6)). Unfortunately it now fails early in the build step. Looks like a libfmt-related error.

```
+ make -j 2 install
[  0%] Generating ../../../magic_mgc.zst.h_
[  1%] Building CXX object tiledb/common/CMakeFiles/baseline.dir/logger.cc.o
In file included from /Users/runner/miniforge3/conda-bld/tiledb_1731294265022/work/tiledb/common/logger.cc:34:
In file included from /Users/runner/miniforge3/conda-bld/tiledb_1731294265022/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/spdlog/fmt/fmt.h:28:
In file included from /Users/runner/miniforge3/conda-bld/tiledb_1731294265022/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/fmt/core.h:5:
In file included from /Users/runner/miniforge3/conda-bld/tiledb_1731294265022/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/fmt/format.h:41:
/Users/runner/miniforge3/conda-bld/tiledb_1731294265022/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/include/fmt/base.h:21:12: fatal error: 'cstddef' file not found
   21 | #  include <cstddef>      // std::byte
      |            ^~~~~~~~~
1 error generated.
make[2]: *** [tiledb/common/CMakeFiles/baseline.dir/build.make:76: tiledb/common/CMakeFiles/baseline.dir/logger.cc.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:4623: tiledb/common/CMakeFiles/baseline.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[  1%] Built target gen_mgc_unarch
make: *** [Makefile:166: all] Error 2
```

h-vetinari commented:
> Looks like a libfmt-related error.

This looks like it's not finding the C++ standard library, which is a pretty catastrophic place to be for a C++ compiler. 🙈 

I don't know why you've set up a wrapper for CXX
https://github.com/conda-forge/tiledb-feedstock/blob/25ae0720c23e34580160a1d3d00584d4c7ba113b/recipe/cxx_wrap.sh#L1-L4
I mean yes, ostensibly to strip `-Werror`, but you could just do that in `build.sh`, or fix it in the source CMake files (since you're already patching things anyway). At least, this wrapper raises my hackles immediately as a potential suspect 😅

h-vetinari commented:
> but you could just do that in `build.sh`

Example (not sure this is all the places where CMake might inject it, but you get the idea):
```bash
  export CFLAGS="${CFLAGS//-Werror/}"
  export CXXFLAGS="${CXXFLAGS//-Werror/}"
```

jdblischak commented:
I rebased this onto main to use the new 2.27.0 (#394), and osx-arm64 successfully built with clang 18! So I cleaned up the PR, including bumping the build number since `cxx_compiler_version` contributes to the binary build hash.

jdblischak commented:
> At least, this wrapper raises my hackles immediately as a potential suspect 

@h-vetinari thanks for your help troubleshooting! Those wrapper scripts also make me uncomfortable. I like your idea of updating the env vars in `build.sh` since it is so much simpler. However, since this PR is currently passing, I would prefer to merge the migration to clang 18 now and troubleshoot the wrapper scripts in a future PR.

conda-forge-admin commented:
Hi! This is the friendly conda-forge automerge bot!

I considered the following status checks when analyzing this PR:
 - **linter**: passed
 - **azure**: passed


Thus the PR was passing and merged! Have a great day!
